File: src/Services/Identity/IdentityServer/ProfileService.fs
module Services

open System.Security.Claims
open System.Threading.Tasks
open IdentityServer4.Extensions
open FSharp.Control.Tasks.V2
open IdentityServer4.Services

type ProfileService(db : Db.IdentityContext) =
    interface IProfileService with
        member this.GetProfileDataAsync(ctx) =
            task {
                let subjectId = ctx.Subject.GetSubjectId()
                let! claimsForUser = UserRepository.getUserClaimsBySubjectId (db, subjectId)
                
                ctx.IssuedClaims <- claimsForUser
                                    |> List.map (fun c -> Claim(c.ClaimType, c.ClaimValue))
                                    |> ResizeArray<Claim>
            } :> Task
            
        member this.IsActiveAsync(ctx) =
            task {
                let subjectId = ctx.Subject.GetSubjectId()
                let! isActive = UserRepository.isUserActive (db, subjectId)
                ctx.IsActive <- isActive
            } :> Task

File: src/Services/Identity/IdentityServer/Services/ProfileService.cs
using System.Linq;
using System.Security.Claims;
using System.Threading.Tasks;
using IdentityServer.Services.Abstractions;
using IdentityServer4.Extensions;
using IdentityServer4.Models;
using IdentityServer4.Services;

namespace IdentityServer.Services
{
    public class ProfileService : IProfileService
    {
        private readonly IIdentityUserRepository _repository;

        public ProfileService(IIdentityUserRepository repository) => 
            _repository = repository;

        public Task GetProfileDataAsync(ProfileDataRequestContext context)
        {
            var subjectId = context.Subject.GetSubjectId();
            var claimsForUser = _repository.GetUserClaimsBySubjectId(subjectId);

            context.IssuedClaims = claimsForUser.Select
                (c => new Claim(c.ClaimType, c.ClaimValue)).ToList();

            return Task.FromResult(0);
        }

        public Task IsActiveAsync(IsActiveContext context)
        {
            var subjectId = context.Subject.GetSubjectId();
            context.IsActive = _repository.IsUserActive(subjectId);

            return Task.FromResult(0);
        }
    }
}